### PR TITLE
C++: Remove uses of abstract from the standard library.

### PR DESCRIPTION
**Pull request by criemen**
Addresses most of https://github.com/github/codeql-c-analysis-team/issues/60

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/Type.qll [hunk 1]
```diff
@@ -577,7 +577,15 @@ class BoolType extends IntegralType {
...
  * ```
  */
-abstract class CharType extends IntegralType { }
+class CharType extends IntegralType {
+  CharType() {
+    builtintypes(underlyingElement(this), _, 5, _, _, _)
+    or
+    builtintypes(underlyingElement(this), _, 6, _, _, _)
+    or
+    builtintypes(underlyingElement(this), _, 7, _, _, _)
```
Comment (MathiasVP) [inline cpp/ql/src/semmle/code/cpp/Type.qll:586]:
```suggestion
    builtintypes(underlyingElement(this), _, [5, 6, 7], _, _, _)
```

File: cpp/ql/src/semmle/code/cpp/internal/QualifiedName.qll [hunk 2]
```diff
@@ -59,7 +59,7 @@ class Namespace extends @namespace {
   }
 }
 
-abstract class Declaration extends @declaration {
+class Declaration extends @declaration {
```
Comment (MathiasVP) [inline cpp/ql/src/semmle/code/cpp/internal/QualifiedName.qll:62]: As this is internal it's not strictly necessary to remove this one, but if it doesn't break anything I don't see why we shouldn't remove it 👍

**Review by MathiasVP — COMMENTED** [review]
Thanks for finishing this!

File: cpp/ql/src/semmle/code/cpp/internal/QualifiedName.qll (hunk 2 above)
Comment (criemen) [reply cpp/ql/src/semmle/code/cpp/internal/QualifiedName.qll:62]: Yeah, I agree, it shouldn't change anything, except that the class doesn't show up in the grep anymore - so why not.

**Review by jbj — COMMENTED** [review]
The changes LGTM, but there's a risk that these changes could have a ripple effect on optimisation of higher-level predicates. Please start a CPP-Differences, comparing the base of the PR to the head of the PR. Instructions are in the team readme.

**Comment by criemen** [comment]
https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1574/

Note: This does not include the changes to the XML parser.

**Comment by criemen** [comment]
https://jenkins-artifacts.internal.semmle.com/static-files/BrG0GVCLQcMtTjRC2DmjxSYaF39V-6Bb363NGrCKQdYxNjA2NDY0MTI3NDAxOjc6Y3JpZW1lbjpqb2IvQ2hhbmdlcy9qb2IvQ1BQLURpZmZlcmVuY2VzLzE1NzQvYXJ0aWZhY3Q=/analysis_time.html means it is looking good, no?

**Review by jbj — APPROVED** [review]

**Pull request merged**